src/Core/MySQL/PacketsProtocolBinary.cpp

```diff
@@ -0,0 +1,363 @@
+                            components.fractional
+                                = DecimalUtils::scaleMultiplier<T>(scale) + (components.whole ? T(-1) : T(1)) * components.fractional;
+                            --components.whole;
+                        }
+                        if (scale > 6)
+                        {
+                            // MySQL Timestamp has max scale of 6
+                            components.fractional /= static_cast<int>(pow(10, scale - 6));
+                        }
+
+                        LocalDateTime ldt = LocalDateTime(components.whole, DateLUT::instance(getDateTimeTimezone(*data_type)));
+
+                        bool has_microseconds = components.fractional != 0;
+                        bool has_time = !(ldt.hour() == 0 && ldt.minute() == 0 && ldt.second() == 0);
+                        if (has_microseconds)
+                        {
+                            payload_size += 12;
+                        }
+                        else if (has_time)
+                        {
+                            payload_size += 8;
+                        }
+                        else
+                        {
+                            payload_size += 5;
+                        }
+                        break;
+                    }
+                    default:
+                        WriteBufferFromOwnString ostr;
+                        serializations[i]->serializeText(*columns[i], row_num, ostr, format_settings);
+                        payload_size += getLengthEncodedStringSize(ostr.str());
+                        serialized[i] = std::move(ostr.str());
+                        break;
+                }
+            }
+        }
+
+        void ResultSetRow::writePayloadImpl(WriteBuffer & buffer) const
+        {
+            buffer.write(static_cast<char>(0x00));
+            buffer.write(null_bitmap.data(), null_bitmap_size);
+            for (size_t i = 0; i < columns.size(); ++i)
+            {
+                ColumnPtr col = columns[i];
+                if (col->isNullAt(row_num))
+                {
+                    continue;
+                }
+
+                DataTypePtr data_type = removeLowCardinality(removeNullable((data_types[i])));
+                TypeIndex type_index = data_type->getTypeId();
+                switch (type_index)
+                {
+                    case TypeIndex::UInt8: {
+                        UInt8 value = assert_cast<const ColumnVector<UInt8> &>(*col).getData()[row_num];
+                        if (data_type->getName() == "Bool")
+                        {
+                            buffer.write(static_cast<char>(1));
+                        }
+                        buffer.write(reinterpret_cast<char *>(&value), 1);
+                        break;
+                    }
+                    case TypeIndex::UInt16: {
+                        UInt16 value = assert_cast<const ColumnVector<UInt16> &>(*col).getData()[row_num];
```

Comment:
We are casting `*col` to the specific column implementation of the corresponding type (e.g. `ColumnVector<UInt16>`). That is okay.

But: This only works if `col` is not of low-cardinality (`ColumnLowCardinality`), or nullable (ColumnNullable), sparse (ColumnSparse) or const (ColumnConst) type. These classes are (possibly chained) wrappers around the actual column structure. In l. 164, we are stripping the low-cardinality and nullness of the *datatype* away, but we would need to do something equivalent for `col`.  Check e.g. `recursiveRemoveSparse()` (Columns/ColumnSparse.h), etc.

Reply:
LowCardinality and Nullable are removed prior to that, so I assumed it should always work as intended.

Reply:
They are only removed from the data type, not from the actual data. I am pretty sure this will go boom, please try it out.